File: pe/visualstudio/PE.FSharp/Problem0012.fs
namespace PE.FSharp.Problem12

open FSharp.MyLibraries.Math

module Problem0012 = 

    let fifteenThousandTriangleNumbers =
        let nthTriangleNumber nth =
            (nth * (nth + 1)) / 2

        [1..15000]
        |> List.map nthTriangleNumber

    let firstTriangleNumberWithMoreThanHowManyFactors n =
        Seq.filter (fun x -> List.length (factors x) > n) fifteenThousandTriangleNumbers
        |> Seq.head

        //Seq evaluates lazily.

File: pe/nonvisualstudio/haskell/Implementation/Problem0012.hs
module Problem0012 where

  import Data.List

  --http://www.mathsisfun.com/algebra/triangular-numbers.html
  nthTriangleNumber nth =
    round
    $(nth * ( nth + 1)) / 2

  infiniteListOfTriangleNumbers = [nthTriangleNumber x | x <- [1..]]

  --http://rosettacode.org/wiki/Factors_of_an_integer#Haskell
  factors n = lows ++ (reverse $ map (div n) lows)
    where lows = filter ((== 0) . mod n) [1..truncate . sqrt $ fromIntegral n]


  firstTriangleNumberWithMoreThanHowManyFactors n =
    head $ filter (\x -> (length $ factors x) > n) infiniteListOfTriangleNumbers
